Add a basic development guide (#83)
PaperMC/docs: docs/paper/dev/getting-started/plugin-yml.md

---
slug: /dev/plugin-yml
---

# Paper Plugin YML

The plugin.yml file is the main configuration file for your plugin. 
It contains information about your plugin, such as its name, version, and description. 
It also contains information about the plugin's dependencies, permissions, and commands.

The plugin.yml file is located in the `resources` directory of your project.
```
src
└── main
    ├── java
    └── resources

>         └── plugin.yml

Reviewer: How about doing like this?
```plaintext
your_plugin_name
└──src
   └── main
       ├── java
       └── resources
           └── plugin.yml
```
This is trivial, though.

Author: Makes sense, have added

```

## Example

Here is an example of a plugin.yml file:

```yaml
name: ExamplePlugin
version: 1.0.0
main: io.papermc.testplugin.ExamplePlugin
description: An example plugin
author: PaperMC
website: https://papermc.io
api-version: 1.19
```

## Fields

:::note

The fields in this section are not in any particular order. 
If they have an asterisk (\*) next to them, that means they are required.

:::

### name*

> The name of your plugin. This is what will be displayed in the plugin list and warning messages.

Reviewer 2: ```suggestion
The name of your plugin. This is what will be displayed in the plugin list and log messages.
```

- `name: ExamplePlugin`

### version*

The current version of the plugin. 
- `version: 1.0.0`

### main*

The main class of your plugin. This is the class that extends `JavaPlugin` and is the entry point to your plugin.
- `main: io.papermc.testplugin.ExamplePlugin`

This is the package path and class name of your main class.

### description

A short description of your plugin and what it does.
- `description: An example plugin`

### author / authors

The author(s) of the plugin. This can be a single author or a list of authors.
- `author: PaperMC`
- `authors: [PaperMC, SpigotMC, Bukkit]`

### website

The website of the plugin. This is useful for linking to a GitHub repository or a plugin page.
- `website: https://papermc.io`

### api-version

> The version of the Paper API that your plugin is using. This doesn't include the minor version. 
> The valid versions are 1.12 - 1.19.

Reviewer 2: ```suggestion
The version of the Paper API that your plugin is using. This doesn't include the minor version. Servers with a version lower than the version specified here will refuse to load the plugin.
The valid versions are 1.13 - 1.19.
```

- `api-version: 1.19`

:::info

If this is not specified, the plugin will be loaded as a legacy plugin and a warning will be printed to the console.

:::

### load

This tells the server when to load the plugin. This can be `STARTUP` or `POSTWORLD`.
- `load: STARTUP`

### prefix

The prefix of the plugin. This is what will be displayed in the log instead of the plugin name.
- `prefix: EpicPaperMCHypePlugin`

### libraries

:::warning

There is debate over whether this is against maven central's TOS. This may be removed in the future.

:::

> This is a list of libraries that your plugin depends on. These libraries will be downloaded from the Maven repository and added to the classpath.

Reviewer 2: ```suggestion
This is a list of libraries that your plugin depends on. These libraries will be downloaded from the Maven central repository and added to the classpath.
```

> This removes the need to shade or relocate the libraries.

Reviewer 2: ```suggestion
This removes the need to shade and relocate the libraries.
```

### description

A short description of your plugin and what it does.
- `description: An example plugin`

### author / authors

The author(s) of the plugin. This can be a single author or a list of authors.
- `author: PaperMC`
- `authors: [PaperMC, SpigotMC, Bukkit]`

### website

The website of the plugin. This is useful for linking to a GitHub repository or a plugin page.
- `website: https://papermc.io`

### api-version

The version of the Paper API that your plugin is using. This doesn't include the minor version.
Servers with a version lower than the version specified here will refuse to load the plugin.
The valid versions are 1.13 - 1.19.
- `api-version: 1.19`

:::info

If this is not specified, the plugin will be loaded as a legacy plugin and a warning will be printed to the console.

:::

### load

This tells the server when to load the plugin. This can be `STARTUP` or `POSTWORLD`.
- `load: STARTUP`

### prefix

The prefix of the plugin. This is what will be displayed in the log instead of the plugin name.
- `prefix: EpicPaperMCHypePlugin`

### libraries

:::warning

There is debate over whether this is against maven central's TOS. This may be removed in the future.

> :::

Reviewer 3: It is against their guidelines, but I don't think there is any plan from either paper or (and I obviously don't know definitively from this one) spigot. If anything, It'd just get moved over to a paper/spigot hosted caching proxy.

  - another.permission.node:
        description: "This is another permission node"
        default: not op
```

The description is the description of the permission node. This is what will be displayed in the permissions list.
The default is the default value of the permission node. This can be `op`/`not op`, or `true`/`false`. This defaults to `op`.
Each permission node can have children. When set to `true` it will inherit the parent permission.

## Commands

This is a list of commands that your plugin uses. This is useful for plugins that use commands to provide features.
```yaml
commands:
  - command:
        description: "This is a command"
        usage: "/command <arg>"
        aliases: [cmd, command]
        permission: permission.node
        permission-message: "You do not have permission to use this command"
```

- The __description__ is the description of the command. This gives a brief description of what the command does.
- The __usage__ is the usage of the command. This is what will be displayed when the player uses `/help <command>`.
- The __aliases__ are a list of aliases that the command can be used with. This is useful for shortening the command.

> - __permission__ is the permission node that the player needs to use the command.

Reviewer 4: maybe add something here that players will not see commands they dont have permission for

Author: Yep, that's something I forgot to be honest 👍

- __permission-message__ is the message that will be displayed when the player does not have permission to use the command.

## Dependencies:

:::note

> If a plugin is specified as a dependency, it will be loaded before your plugin.

Reviewer 3: I would add a note here about dependency/loadbefore loops. In the future, paper will warn about these as they cause indeterminate behaviour and can break plugins. See https://github.com/PaperMC/Paper/issues/5992 as well as a number of other similar issues.
